File: sdk/consensus/slashing.py
"""
Slashing Module for Validator Penalties.

Matches luxtensor-consensus/src/slashing.rs.

Handles:
- Offline validators (missed blocks)
- Double signing detection
- Invalid block/weight proposals
- Jailing mechanism
"""

from dataclasses import dataclass, field
from enum import Enum
from typing import Dict, List, Optional, Tuple
from decimal import Decimal
import time


class SlashReason(str, Enum):
    """Reasons for slashing a validator."""
    OFFLINE = "offline"              # Missed too many blocks
    DOUBLE_SIGNING = "double_signing"  # Signed two blocks at same height
    INVALID_BLOCK = "invalid_block"   # Proposed invalid block
    INVALID_WEIGHTS = "invalid_weights"  # Submitted invalid weights
    CUSTOM = "custom"                 # Custom reason

    def slash_percent(self, config: 'SlashingConfig') -> int:
        """Get slash percentage for this reason."""
        return {
            SlashReason.OFFLINE: config.offline_slash_percent,
            SlashReason.DOUBLE_SIGNING: config.double_sign_slash_percent,
            SlashReason.INVALID_BLOCK: config.invalid_block_slash_percent,
            SlashReason.INVALID_WEIGHTS: config.invalid_weights_slash_percent,
            SlashReason.CUSTOM: 1,
        }[self]


@dataclass
class SlashingConfig:
    """
    Slashing configuration.

    Matches luxtensor-consensus SlashingConfig.
    """
    # Slash percentages
    offline_slash_percent: int = 1          # 1% for being offline
    double_sign_slash_percent: int = 5      # 5% for double signing
    invalid_block_slash_percent: int = 2    # 2% for invalid block
    invalid_weights_slash_percent: int = 1  # 1% for invalid weights

    # Thresholds
    max_missed_blocks: int = 100            # Blocks before offline slash
    signed_blocks_window: int = 1000        # Window for tracking missed blocks

    # Jail duration
    jail_duration: int = 10000              # Blocks of jail time

    @classmethod
    def default(cls) -> 'SlashingConfig':
        return cls()


@dataclass
class SlashingEvidence:
    """
    Evidence of misbehavior.

    Matches luxtensor-consensus SlashingEvidence.
    """
    validator: str          # Validator address
    reason: SlashReason     # Reason for slashing
    height: int             # Block height when detected
    evidence_hash: str      # Hash of evidence data
    timestamp: int = 0      # When detected
    details: str = ""       # Additional details

    def __post_init__(self):
        if self.timestamp == 0:
            self.timestamp = int(time.time())


@dataclass
class SlashEvent:
    """
    Slashing event record.

    Matches luxtensor-consensus SlashEvent.
    """
    validator: str              # Validator address
    reason: SlashReason         # Reason
    slash_percent: int          # Percentage slashed
    amount_slashed: int         # Amount slashed (in base units)
    height: int                 # Block height
    timestamp: int              # Unix timestamp
    jailed: bool = False        # Whether validator was jailed
    jail_until: int = 0         # Block height to unjail

    def to_dict(self) -> Dict:
        return {
            "validator": self.validator,
            "reason": self.reason.value,
            "slash_percent": self.slash_percent,
            "amount_slashed": str(self.amount_slashed),
            "amount_slashed_mdt": str(Decimal(self.amount_slashed) / Decimal(10**18)),
            "height": self.height,
            "timestamp": self.timestamp,
            "jailed": self.jailed,
            "jail_until": self.jail_until if self.jailed else None,
        }


@dataclass
class JailStatus:
    """
    Jail status for a validator.

    Matches luxtensor-consensus JailStatus.
    """
    validator: str
    jailed_at: int          # Block height when jailed
    jail_until: int         # Block height when can unjail
    reason: SlashReason     # Reason for jailing
    slash_count: int = 1    # Number of times slashed


class SlashingManager:
    """
    Slashing Manager - handles validator penalties.

    Matches luxtensor-consensus SlashingManager.
    """

    def __init__(self, config: Optional[SlashingConfig] = None):
        self.config = config or SlashingConfig.default()
        self._missed_blocks: Dict[str, int] = {}  # validator -> count
        self._slash_history: Dict[str, List[SlashEvent]] = {}  # validator -> events
        self._jail_status: Dict[str, JailStatus] = {}  # validator -> status
        self._block_signatures: Dict[int, Dict[str, List[str]]] = {}  # height -> validator -> signatures

    def record_missed_block(self, validator: str):
        """Record a missed block for validator."""
        validator = validator.lower()
        self._missed_blocks[validator] = self._missed_blocks.get(validator, 0) + 1

    def reset_missed_blocks(self, validator: str):
        """Reset missed blocks for validator (called when they produce a block)."""
        self._missed_blocks[validator.lower()] = 0

    def check_offline(self, validator: str) -> Optional[SlashingEvidence]:
        """Check if validator should be slashed for being offline."""
        validator = validator.lower()
        missed = self._missed_blocks.get(validator, 0)

        if missed >= self.config.max_missed_blocks:
            return SlashingEvidence(
                validator=validator,
                reason=SlashReason.OFFLINE,
                height=0,  # Will be set by caller
                evidence_hash=f"offline_{validator}_{missed}",
                details=f"Missed {missed} blocks",
            )
        return None

    def record_block_signature(
        self,
        height: int,
        block_hash: str,
        validator: str,
        signature_hash: str,
    ):
        """Record potential double signing."""
        validator = validator.lower()
        if height not in self._block_signatures:
            self._block_signatures[height] = {}
        if validator not in self._block_signatures[height]:
            self._block_signatures[height][validator] = []

        self._block_signatures[height][validator].append(f"{block_hash}:{signature_hash}")

    def check_double_signing(self, height: int) -> List[SlashingEvidence]:
        """Check for double signing at a height."""
        evidence = []

        sigs = self._block_signatures.get(height, {})
        for validator, signatures in sigs.items():
            # Check if validator signed multiple different blocks
            unique_blocks = set(s.split(":")[0] for s in signatures)
            if len(unique_blocks) > 1:
                evidence.append(SlashingEvidence(
                    validator=validator,
                    reason=SlashReason.DOUBLE_SIGNING,
                    height=height,
                    evidence_hash=f"double_sign_{validator}_{height}",
                    details=f"Signed {len(unique_blocks)} different blocks at height {height}",
                ))

        return evidence

    def slash(
        self,
        evidence: SlashingEvidence,
        current_height: int,
        validator_stake: int,
    ) -> SlashEvent:
        """
        Execute slash on a validator.

        Args:
            evidence: Evidence of misbehavior
            current_height: Current block height
            validator_stake: Validator's current stake

        Returns:
            SlashEvent with details
        """
        validator = evidence.validator.lower()
        slash_percent = evidence.reason.slash_percent(self.config)

        # Calculate slashed amount
        amount_slashed = (validator_stake * slash_percent) // 100

        # Check if should be jailed
        should_jail = evidence.reason in [SlashReason.DOUBLE_SIGNING, SlashReason.OFFLINE]
        jail_until = current_height + self.config.jail_duration if should_jail else 0

        # Create event
        event = SlashEvent(
            validator=validator,
            reason=evidence.reason,
            slash_percent=slash_percent,
            amount_slashed=amount_slashed,
            height=current_height,
            timestamp=int(time.time()),
            jailed=should_jail,
            jail_until=jail_until,
        )

        # Record in history
        if validator not in self._slash_history:
            self._slash_history[validator] = []
        self._slash_history[validator].append(event)

        # Update jail status
        if should_jail:
            self._jail_status[validator] = JailStatus(
                validator=validator,
                jailed_at=current_height,
                jail_until=jail_until,
                reason=evidence.reason,
                slash_count=len(self._slash_history[validator]),
            )

        # Reset missed blocks
        self.reset_missed_blocks(validator)

        return event

    def is_jailed(self, validator: str) -> bool:
        """Check if validator is jailed."""
        return validator.lower() in self._jail_status

    def get_jail_status(self, validator: str) -> Optional[JailStatus]:
        """Get jail status."""
        return self._jail_status.get(validator.lower())

    def process_unjail(self, current_height: int) -> List[str]:
        """Process unjailing (called each block). Returns list of unjailed validators."""
        unjailed = []

        for validator, status in list(self._jail_status.items()):
            if current_height >= status.jail_until:
                del self._jail_status[validator]
                unjailed.append(validator)

        return unjailed

    def get_slash_history(self, validator: str) -> List[SlashEvent]:
        """Get slash history for an address."""
        return self._slash_history.get(validator.lower(), [])

    def get_all_slash_events(self) -> List[SlashEvent]:
        """Get all slash events."""
        all_events = []
        for events in self._slash_history.values():
            all_events.extend(events)
        return sorted(all_events, key=lambda e: e.height, reverse=True)

    def get_total_slashed(self, validator: str) -> int:
        """Get total slashed amount for a validator."""
        events = self.get_slash_history(validator)
        return sum(e.amount_slashed for e in events)

    def cleanup_old_signatures(self, current_height: int, max_age: int = 1000):
        """Clean up old signatures (call periodically)."""
        min_height = current_height - max_age
        self._block_signatures = {
            h: sigs for h, sigs in self._block_signatures.items()
            if h >= min_height
        }


# Module exports
__all__ = [
    "SlashReason",
    "SlashingConfig",
    "SlashingEvidence",
    "SlashEvent",
    "JailStatus",
    "SlashingManager",
]

File: luxtensor/crates/luxtensor-consensus/src/slashing.rs
// Slashing module for validator penalties
// Handles: offline validators, double signing, malicious behavior

use crate::error::ConsensusError;
use crate::validator::ValidatorSet;
use luxtensor_core::types::{Address, Hash};
use parking_lot::RwLock;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;
use tracing::{info, warn};

/// Slashing configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SlashingConfig {
    /// Blocks before considered offline
    pub offline_threshold: u64,
    /// Percentage to slash for being offline (0-100)
    pub offline_slash_percent: u8,
    /// Percentage to slash for double signing (0-100)
    pub double_sign_slash_percent: u8,
    /// Percentage to slash for invalid block (0-100)
    pub invalid_block_slash_percent: u8,
    /// Minimum slash amount (absolute)
    pub min_slash_amount: u128,
    /// Jail duration in blocks
    pub jail_duration: u64,
}

impl Default for SlashingConfig {
    fn default() -> Self {
        Self {
            offline_threshold: 100,                      // 100 missed blocks
            offline_slash_percent: 1,                    // 1% stake
            double_sign_slash_percent: 10,               // 10% stake
            invalid_block_slash_percent: 5,              // 5% stake
            min_slash_amount: 1_000_000_000_000_000_000, // 1 token
            jail_duration: 7200,                         // ~24 hours @ 12s blocks
        }
    }
}

/// Reasons for slashing
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum SlashReason {
    /// Validator missed too many blocks
    Offline,
    /// Validator signed two different blocks at same height
    DoubleSigning,
    /// Validator proposed invalid block
    InvalidBlock,
    /// Validator submitted invalid weights
    InvalidWeights,
    /// Custom reason with percentage
    Custom(u8),
}

impl SlashReason {
    /// Get slash percentage for this reason
    pub fn slash_percent(&self, config: &SlashingConfig) -> u8 {
        match self {
            SlashReason::Offline => config.offline_slash_percent,
            SlashReason::DoubleSigning => config.double_sign_slash_percent,
            SlashReason::InvalidBlock => config.invalid_block_slash_percent,
            SlashReason::InvalidWeights => config.offline_slash_percent,
            SlashReason::Custom(pct) => *pct,
        }
    }
}

/// Evidence of misbehavior
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SlashingEvidence {
    /// Validator address
    pub validator: Address,
    /// Reason for slash
    pub reason: SlashReason,
    /// Block height when offense occurred
    pub height: u64,
    /// Evidence hash (block signatures, etc.)
    pub evidence_hash: Option<Hash>,
    /// Timestamp
    pub timestamp: u64,
}

/// Slashing event record
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SlashEvent {
    /// Evidence that triggered slash
    pub evidence: SlashingEvidence,
    /// Amount slashed
    pub amount_slashed: u128,
    /// Whether validator was jailed
    pub jailed: bool,
    /// Block height of slash execution
    pub executed_at: u64,
}

/// Jail status for a validator
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JailStatus {
    /// When jailed
    pub jailed_at: u64,
    /// When jail ends
    pub release_at: u64,
    /// Reason for jailing
    pub reason: SlashReason,
}

/// Maximum entries in double-sign evidence map before pruning old entries.
/// Prevents unbounded memory growth from malicious evidence submission.
const MAX_EVIDENCE_ENTRIES: usize = 10_000;

/// Maximum slash events retained in history.
const MAX_SLASH_HISTORY: usize = 50_000;

/// Slashing manager
pub struct SlashingManager {
    config: SlashingConfig,

    /// Missed block counter per validator
    missed_blocks: RwLock<HashMap<Address, u64>>,

    /// Double signing evidence: (height, block_hash) -> signatures
    double_sign_evidence: RwLock<HashMap<(u64, Hash), Vec<(Address, Hash)>>>,

    /// Jailed validators
    jailed: RwLock<HashMap<Address, JailStatus>>,

    /// Slash history (bounded — oldest entries pruned when exceeding MAX_SLASH_HISTORY)
    slash_history: RwLock<Vec<SlashEvent>>,

    /// Reference to validator set
    validator_set: Arc<RwLock<ValidatorSet>>,
}

impl SlashingManager {
    /// Create new slashing manager
    pub fn new(config: SlashingConfig, validator_set: Arc<RwLock<ValidatorSet>>) -> Self {
        Self {
            config,
            missed_blocks: RwLock::new(HashMap::new()),
            double_sign_evidence: RwLock::new(HashMap::new()),
            jailed: RwLock::new(HashMap::new()),
            slash_history: RwLock::new(Vec::new()),
            validator_set,
        }
    }

    /// Record a missed block for validator
    pub fn record_missed_block(&self, validator: &Address) {
        let mut missed = self.missed_blocks.write();
        let count = missed.entry(*validator).or_insert(0);
        *count += 1;

        if *count >= self.config.offline_threshold {
            warn!("Validator {:?} missed {} blocks, threshold reached", validator, count);
        }
    }

    /// Reset missed blocks for validator (called when they produce a block)
    pub fn reset_missed_blocks(&self, validator: &Address) {
        self.missed_blocks.write().remove(validator);
    }

    /// Check if validator should be slashed for being offline
    ///
    /// `current_height` is used as the deterministic timestamp for consensus.
    /// SECURITY: Previously used SystemTime::now() which is non-deterministic
    /// and could cause consensus disagreements between nodes.
    pub fn check_offline(
        &self,
        validator: &Address,
        current_height: u64,
    ) -> Option<SlashingEvidence> {
        let missed = self.missed_blocks.read();
        if let Some(&count) = missed.get(validator) {
            if count >= self.config.offline_threshold {
                return Some(SlashingEvidence {
                    validator: *validator,
                    reason: SlashReason::Offline,
                    height: current_height,
                    evidence_hash: None,
                    timestamp: current_height, // Deterministic: use block height as timestamp
                });
            }
        }
        None
    }

    /// Record potential double signing
    pub fn record_block_signature(
        &self,
        height: u64,
        block_hash: Hash,
        validator: Address,
        signature_hash: Hash,
    ) {
        let mut evidence = self.double_sign_evidence.write();

        // SECURITY: Prune oldest entries if evidence map exceeds bounds.
        // An attacker could spam evidence entries to cause OOM.
        if evidence.len() >= MAX_EVIDENCE_ENTRIES {
            // Remove entries with the lowest heights (oldest)
            let mut heights: Vec<u64> = evidence.keys().map(|(h, _)| *h).collect();
            heights.sort_unstable();
            let cutoff = heights.get(heights.len() / 2).copied().unwrap_or(0);
            evidence.retain(|(h, _), _| *h > cutoff);
            warn!(
                "Double-sign evidence pruned: removed entries at height <= {}, {} remaining",
                cutoff,
                evidence.len()
            );
        }

        let key = (height, block_hash);
        let sigs = evidence.entry(key).or_insert_with(Vec::new);
        sigs.push((validator, signature_hash));
    }

    /// Check for double signing at a height
    pub fn check_double_signing(&self, height: u64) -> Vec<SlashingEvidence> {
        let evidence = self.double_sign_evidence.read();
        let mut offenders = Vec::new();

        // Find validators who signed multiple blocks at same height
        let mut validator_blocks: HashMap<Address, Vec<Hash>> = HashMap::new();

        for ((h, block_hash), sigs) in evidence.iter() {
            if *h == height {
                for (validator, _sig_hash) in sigs {
                    validator_blocks.entry(*validator).or_insert_with(Vec::new).push(*block_hash);
                }
            }
        }

        for (validator, blocks) in validator_blocks {
            if blocks.len() > 1 {
                // Same validator signed different blocks at same height!
                warn!(
                    "Double signing detected! Validator {:?} signed {} blocks at height {}",
                    validator,
                    blocks.len(),
                    height
                );

                offenders.push(SlashingEvidence {
                    validator,
                    reason: SlashReason::DoubleSigning,
                    height,
                    evidence_hash: Some(blocks[0]), // First conflicting block
                    timestamp: height,              // Deterministic: use block height as timestamp
                });
            }
        }

        offenders
    }

    /// Execute slash on a validator
    ///
    /// LOCK ORDER: validator_set → jailed → slash_history (matches process_unjail() to prevent ABBA deadlock)
    pub fn slash(
        &self,
        evidence: SlashingEvidence,
        current_height: u64,
    ) -> Result<SlashEvent, ConsensusError> {
        let percent = evidence.reason.slash_percent(&self.config);

        // Get validator stake
        let mut validator_set = self.validator_set.write();
        let validator = validator_set.get_validator(&evidence.validator).ok_or_else(|| {
            ConsensusError::ValidatorNotFound(format!("{:?}", evidence.validator))
        })?;

        // 🔧 FIX: Handle zero-stake validators — still jail them even if
        // monetary slash is zero (prevents unpunished misbehaviour)
        let slash_amount = if validator.stake == 0 {
            warn!(
                "Validator {:?} has zero stake — monetary slash skipped but jail applies",
                evidence.validator
            );
            0
        } else {
            let amount = (validator.stake as u128 * percent as u128) / 100;
            let amount = amount.max(self.config.min_slash_amount);
            amount.min(validator.stake) // Can't slash more than stake
        };

        // Apply slash (only if amount > 0)
        if slash_amount > 0 {
            validator_set
                .slash_stake(&evidence.validator, slash_amount)
                .map_err(|e| ConsensusError::SlashingFailed(e.to_string()))?;
        }

        // Jail validator for serious offenses
        let should_jail =
            matches!(evidence.reason, SlashReason::DoubleSigning | SlashReason::InvalidBlock);

        if should_jail {
            let jail_status = JailStatus {
                jailed_at: current_height,
                release_at: current_height + self.config.jail_duration,
                reason: evidence.reason,
            };
            self.jailed.write().insert(evidence.validator, jail_status);

            // Deactivate validator
            if let Err(e) = validator_set.deactivate_validator(&evidence.validator) {
                warn!("Failed to deactivate jailed validator: {}", e);
            }
        }

        // Record event
        let event = SlashEvent {
            evidence,
            amount_slashed: slash_amount,
            jailed: should_jail,
            executed_at: current_height,
        };

        self.slash_history.write().push(event.clone());

        // SECURITY: Prune oldest slash history entries to prevent unbounded growth
        {
            let mut history = self.slash_history.write();
            if history.len() > MAX_SLASH_HISTORY {
                let drain_count = history.len() - MAX_SLASH_HISTORY;
                history.drain(..drain_count);
            }
        }

        info!(
            "Slashed validator {:?}: {} tokens, jailed: {}",
            event.evidence.validator, slash_amount, should_jail
        );

        // Reset missed blocks counter
        self.missed_blocks.write().remove(&event.evidence.validator);

        Ok(event)
    }

    /// Check if validator is jailed
    pub fn is_jailed(&self, validator: &Address) -> bool {
        self.jailed.read().contains_key(validator)
    }

    /// Get jail status
    pub fn get_jail_status(&self, validator: &Address) -> Option<JailStatus> {
        self.jailed.read().get(validator).cloned()
    }

    /// Process unjailing (called each block)
    ///
    /// LOCK ORDER: validator_set → jailed (matches slash() to prevent ABBA deadlock)
    pub fn process_unjail(&self, current_height: u64) -> Vec<Address> {
        let mut validator_set = self.validator_set.write();
        let mut jailed = self.jailed.write();

        let mut unjailed = Vec::new();

        jailed.retain(|validator, status| {
            if current_height >= status.release_at {
                info!("Unjailing validator {:?} at height {}", validator, current_height);

                // Reactivate validator
                if let Err(e) = validator_set.activate_validator(validator) {
                    warn!("Failed to reactivate unjailed validator: {}", e);
                }

                unjailed.push(*validator);
                false // Remove from jailed map
            } else {
                true // Keep in jailed map
            }
        });

        unjailed
    }

    /// Get slash history for an address
    pub fn get_slash_history(&self, validator: &Address) -> Vec<SlashEvent> {
        self.slash_history
            .read()
            .iter()
            .filter(|e| &e.evidence.validator == validator)
            .cloned()
            .collect()
    }

    /// Get all slash events
    pub fn get_all_slash_events(&self) -> Vec<SlashEvent> {
        self.slash_history.read().clone()
    }

    /// Get total slashed amount for a validator
    pub fn get_total_slashed(&self, validator: &Address) -> u128 {
        self.slash_history
            .read()
            .iter()
            .filter(|e| &e.evidence.validator == validator)
            .map(|e| e.amount_slashed)
            .sum()
    }

    /// Clean up old evidence (call periodically)
    pub fn cleanup_old_evidence(&self, current_height: u64, max_age: u64) {
        let cutoff = current_height.saturating_sub(max_age);

        self.double_sign_evidence.write().retain(|(height, _), _| *height >= cutoff);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::validator::Validator;

    fn setup_test() -> (SlashingManager, Address) {
        let validator_set = Arc::new(RwLock::new(ValidatorSet::new()));

        // Add a test validator
        let validator = Validator {
            address: Address::zero(),
            stake: 100_000_000_000_000_000_000u128, // 100 tokens
            public_key: [0u8; 32],
            active: true,
            rewards: 0,
            last_active_slot: 0,
            activation_epoch: 0,
        };
        validator_set.write().add_validator(validator.clone()).unwrap();

        let manager = SlashingManager::new(SlashingConfig::default(), validator_set);

        (manager, Address::zero())
    }

    #[test]
    fn test_missed_blocks_tracking() {
        let (manager, validator) = setup_test();

        // Record missed blocks
        for _ in 0..50 {
            manager.record_missed_block(&validator);
        }

        // Should not trigger slash yet (threshold is 100)
        assert!(manager.check_offline(&validator, 50).is_none());

        // Record more
        for _ in 0..50 {
            manager.record_missed_block(&validator);
        }

        // Now should trigger
        assert!(manager.check_offline(&validator, 100).is_some());
    }

    #[test]
    fn test_slash_execution() {
        let (manager, validator) = setup_test();

        let evidence = SlashingEvidence {
            validator,
            reason: SlashReason::Offline,
            height: 100,
            evidence_hash: None,
            timestamp: 0,
        };

        let event = manager.slash(evidence, 100).unwrap();

        // Should have slashed 1% of 100 tokens = 1 token
        assert!(event.amount_slashed >= 1_000_000_000_000_000_000);
        assert!(!event.jailed); // Offline doesn't jail
    }

    #[test]
    fn test_double_signing_jails() {
        let (manager, validator) = setup_test();

        let evidence = SlashingEvidence {
            validator,
            reason: SlashReason::DoubleSigning,
            height: 100,
            evidence_hash: Some([1u8; 32]),
            timestamp: 0,
        };

        let event = manager.slash(evidence, 100).unwrap();

        // Should be jailed
        assert!(event.jailed);
        assert!(manager.is_jailed(&validator));
    }

    #[test]
    fn test_unjailing() {
        let (manager, validator) = setup_test();

        // Jail the validator
        let evidence = SlashingEvidence {
            validator,
            reason: SlashReason::DoubleSigning,
            height: 100,
            evidence_hash: None,
            timestamp: 0,
        };
        manager.slash(evidence, 100).unwrap();

        // Process before jail ends - should stay jailed
        let unjailed = manager.process_unjail(100 + 1000);
        assert!(unjailed.is_empty());

        // Process after jail ends
        let unjailed = manager.process_unjail(100 + 7200 + 1);
        assert_eq!(unjailed.len(), 1);
        assert_eq!(unjailed[0], validator);
        assert!(!manager.is_jailed(&validator));
    }
}
